### PR TITLE
Simple autorebuild mechanism for hosted apps

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
In the 0.3.0 release we had to disable autorebuild for hosted apps due to a bug. This PR re-enables it, replacing the implementation with a much simpler alternative.

This implementation should be understood as a stopgap measure until some future version of ASP.NET Core and VS can hopefully extend the native autorebuild support to rebuild when there are changes in referenced projects. Once (and if) that is implemented, both the standalone and hosted variants of Blazor's autorebuild feature could be removed because both would inherit it from ASP.NET Core.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs [hunk 1]
```diff
@@ -19,7 +21,37 @@ internal static class AutoRebuildExtensions
...
+                        var fileToTouch = FindFirstFileInDirectoryRecursive(hostAppContentRootPath, "*.cs");
+                        if (!string.IsNullOrEmpty(fileToTouch))
+                        {
+                            File.SetLastWriteTime(fileToTouch, DateTime.Now);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // If we don't have permission to write these files, autorebuild will not be enabled
+                        var loggerFactory = (ILoggerFactory)appBuilder.ApplicationServices.GetService(typeof(ILoggerFactory));
```
Comment (attilah) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Why not using the generic versions of the DI methods?

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs [hunk 2]
```diff
@@ -77,6 +109,12 @@ public static void UseAutoRebuild(this IApplicationBuilder appBuilder, BlazorCon
             });
         }
 
+        private static string FindFirstFileInDirectoryRecursive(string rootDirectory, string pattern)
```
Comment (attilah) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: What do you think of leveraging the new .Net Core 2.1 FileSystemEnumerable, that's more lightweight and offers plenty of options.

Something similar:

```
            var enumerationOptions = new EnumerationOptions
            {
                IgnoreInaccessible = true,
                RecurseSubdirectories = true
            };

            var fileToTouch = new FileSystemEnumerable<string>(hostAppContentRootPath, (ref FileSystemEntry entry) => entry.ToFullPath(), enumerationOptions)
            {
                ShouldIncludePredicate = (ref FileSystemEntry entry) => !entry.IsDirectory && entry.FileName.EndsWith(".cs", StringComparison.OrdinalIgnoreCase)
            };
 ```

Comment (rynowak) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Is https://msdn.microsoft.com/en-us/library/dd383571(v=vs.110).aspx a suitable replacement for this or is something non-obvious happening here?

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs [hunk 3]
```diff
@@ -19,7 +21,37 @@ internal static class AutoRebuildExtensions
...
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // If we don't have permission to write these files, autorebuild will not be enabled
+                        var loggerFactory = (ILoggerFactory)appBuilder.ApplicationServices.GetService(typeof(ILoggerFactory));
+                        var logger = loggerFactory?.CreateLogger(typeof (AutoRebuildExtensions));
+                        logger?.LogInformation(ex,
+                            "Cannot autorebuild because there was an error when writing to a file in '{0}'.",
+                            hostAppContentRootPath);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Should this be a warning?

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs [hunk 4]
```diff
@@ -19,7 +21,37 @@ internal static class AutoRebuildExtensions
...
+                        if (!string.IsNullOrEmpty(fileToTouch))
+                        {
+                            File.SetLastWriteTime(fileToTouch, DateTime.Now);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // If we don't have permission to write these files, autorebuild will not be enabled
+                        var loggerFactory = (ILoggerFactory)appBuilder.ApplicationServices.GetService(typeof(ILoggerFactory));
+                        var logger = loggerFactory?.CreateLogger(typeof (AutoRebuildExtensions));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Is it actually possible for DI to not be registered? I think it's baked in at the lowest possible level in hosted ASP.NET Core.

You could also shorten this to `appBuilder.ApplicationServices.GetRequiredService<ILogger<AutoRebuildExtensions>>()` if you so desire

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 2 above)
Comment (attilah) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: I've checked the implementation and now ```Directory``` is backed by FileSystemEntry* stuff, which is really cool, the whole recursive file search could be a single liner with (static) options:

```
var enumerationOptions = new EnumerationOptions
{
    IgnoreInaccessible = true,
    RecurseSubdirectories = true
};

var fileToTouch = Directory.GetFiles(hostAppContentRootPath, "*.cs", enumerationOptions).FirstOrDefault();
```

Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: I would have thought that `RecurseSubdirectories = true` would force it to build a list containing every single `.cs` file in the entire directory hierarchy, of which there might be many. I was trying to let to stop as soon as it finds any directory containing any `.cs` file.

Comment (attilah) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Fortunately not, if you look at the implementation the same API is used with a different predicate:

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System/IO/Directory.cs#L160

Comment (rynowak) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: From EnumerateFiles remarks: 

> The EnumerateFiles and GetFiles methods differ as follows: When you use EnumerateFiles, you can start enumerating the collection of names before the whole collection is returned; when you use GetFiles, you must wait for the whole array of names to be returned before you can access the array. Therefore, when you are working with many files and directories, EnumerateFiles can be more efficient.

Comment (attilah) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Yup, that's true, but the single liner above by changing ```GetFiles``` to ```EnumerateFiles``` would still remove the need for Linq and others.

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs [hunk 5]
```diff
@@ -19,7 +21,37 @@ internal static class AutoRebuildExtensions
         private static string[] _includedSuffixes = new[] { ".cs", ".cshtml", "index.html" };
         private static string[] _excludedDirectories = new[] { "obj", "bin" };
 
-        public static void UseAutoRebuild(this IApplicationBuilder appBuilder, BlazorConfig config)
+        public static void UseHostedAutoRebuild(this IApplicationBuilder appBuilder, BlazorConfig config, string hostAppContentRootPath)
+        {
+            var isFirstFileWrite = true;
+            WatchFileSystem(config, () =>
```
Comment (MisinformedDNA) [inline src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: You'll have to excuse my ignorance, but won't `FileSystemWatcher` be GCed?

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Done

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Good point - done.

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: `Directory.EnumerateFiles` is a good fit here, so have switched to that. Thanks for the tip.

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Yes, that would be reasonable. Done.

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 2 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: I'm sure that would work fine, but `Directory.EnumerateFiles` is even simpler and given the use case (invoked at most once per app build) there is no chance that any perf difference would be observable.

File: src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs]: Good point! Now accounted for.
